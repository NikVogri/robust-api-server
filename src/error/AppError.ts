interface IAppError {
    message: string;
    statusCode: number;
    thrownError?: Error | unknown;
    payload?: Record<string, unknown>;
}

export class AppError extends Error {
    thrownError?: Error | unknown;
    statusCode: number;
    payload?: Record<string, unknown>;

    constructor(public data: IAppError) {
        super(data.message);

        this.thrownError = data.thrownError;
        this.statusCode = data.statusCode;
        this.payload = data.payload;
    }

    static toStringifiableObject(error: Error | unknown) {
        if (!(error instanceof Error)) return {};

        const err: Record<string, unknown> = {};

        Object.getOwnPropertyNames(error).forEach((prop) => {
            if ((error[prop as keyof Error] as unknown) instanceof Error) {
                err[prop] = AppError.toStringifiableObject(error);
            } else err[prop] = error[prop as keyof Error];
        });

        return err;
    }
}
