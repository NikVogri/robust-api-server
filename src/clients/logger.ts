import { Request, Response } from 'express';
import { Service } from 'typedi';
import winston from 'winston';
import { AppError } from '../error/AppError';

@Service()
export class Logger {
    private logger: winston.Logger;

    private custom = {
        levels: {
            error: 0,
            info: 2,
            http: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.logger = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'http',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'info',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'error',
                format: winston.format.combine(
                    // winston.format.timestamp(), TODO: add timestamp
                    // TODO: make sure error only gets logged once! remove default transports for error?
                    winston.format.json(),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error, metadata?: Record<string, unknown>) {
        this.logger.error(message, AppError.toStringifiableObject(error), metadata);
    }

    info(message: string, metadata?: Record<string, unknown>) {
        this.logger.info(message, metadata);
    }

    http(req: Request, res: Response, error?: AppError) {
        const log = this.buildHttpLog(req, res, error);
        this.logger.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: AppError): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            url: req.url,
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: req.query,
                body: req.body ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: {
                message: error?.message,
                stack: error?.stack,
                thrown: error?.thrownError ? AppError.toStringifiableObject(error.thrownError) : undefined,
            },
        });
    }
}
