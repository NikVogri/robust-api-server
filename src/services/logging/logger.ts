import winston from 'winston';
import { AppError } from '../../error/AppError';
import { singleton } from 'tsyringe';
import { Request, Response } from '../../types';

@singleton()
export class Logger {
    public readonly winston: winston.Logger;

    private readonly custom = {
        levels: {
            error: 0,
            info: 1,
            http: 2,
            verbose: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
            verbose: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.winston = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'verbose',
                format: winston.format.combine(
                    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
                    winston.format.printf(
                        (i) => `${i.level} ${i.timestamp}: ${i.message} ${JSON.stringify(i.stack) ?? ''}`,
                    ),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error) {
        this.winston.error(message, AppError.toStringifiableObject(error));
    }

    info(message: string) {
        this.winston.info(message);
    }

    http(req: Request, res: Response, error?: Error) {
        const log = this.buildHttpLog(req, res, error);
        this.winston.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: unknown): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: Object.keys(req.query).length > 0 ? req.query : undefined,
                params: Object.keys(req.params).length > 0 ? req.params : undefined,
                body: Object.keys(req.body).length > 0 ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: this.buildErrorObject(error),
        });
    }

    private buildErrorObject(error?: unknown) {
        if (!error || !(error instanceof Error)) return undefined;

        const output: Record<string, unknown> = {
            message: error?.message,
        };

        if (error instanceof AppError && error.thrownError) {
            output['thrown'] = AppError.toStringifiableObject(error.thrownError);
        }

        return output;
    }
}
