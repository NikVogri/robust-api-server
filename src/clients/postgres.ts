import { Pool } from 'pg';
import { Logger } from './logger';
import { singleton } from 'tsyringe';

@singleton()
export class Postgres {
    readonly pool: Pool;

    constructor(private logger: Logger) {
        this.pool = new Pool({
            connectionString: process.env.PG_DB_CONNECTION_STR,
        });
    }

    async testConnection(): Promise<void> {
        await this.pool.query('SELECT 1 + 1;');
        this.logger.info('Successfully connected to Postgres');
    }

    async getClient() {
        return await this.pool.connect();
    }
}
