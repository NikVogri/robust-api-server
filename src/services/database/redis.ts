import { Redis as IORedis } from 'ioredis';
import { Logger } from '../logging/logger';
import { singleton } from 'tsyringe';

@singleton()
export class Redis {
    readonly client: IORedis;

    constructor(private logger: Logger) {
        this.client = new IORedis(process.env.REDIS_DB_CONNECTION_STR!, {
            maxRetriesPerRequest: null,
        });
    }

    async testConnection(): Promise<void> {
        await this.client.ping();
        this.logger.info('Successfully connected to Redis');
    }
}
