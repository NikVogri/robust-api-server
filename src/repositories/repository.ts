import { AppError } from '../error/AppError';
import { Postgres } from '../services/database/postgres';

export abstract class Repository<T> {
    constructor(
        protected table: string,
        protected pg: Postgres,
    ) {}

    async getAll(): Promise<T[]> {
        try {
            const { rows } = await this.pg.pool.query(`SELECT * FROM ${this.table}`);
            return rows;
        } catch (err) {
            throw new AppError({
                message: 'Unable to get all cats',
                statusCode: 500,
                thrownError: err,
            });
        }
    }

    async getById(id: string): Promise<T | undefined> {
        try {
            const { rows } = await this.pg.pool.query(`SELECT * FROM ${this.table} WHERE id = $1`, [id]);
            return rows[0];
        } catch (err) {
            throw new AppError({
                message: 'Unable to get cat by id',
                statusCode: 500,
                thrownError: err,
            });
        }
    }

    async delete(id: string): Promise<void> {
        try {
            await this.pg.pool.query(`DELETE FROM ${this.table} WHERE id = $1`, [id]);
        } catch (err) {
            throw new AppError({
                message: 'Unable to delete cat',
                statusCode: 500,
                thrownError: err,
            });
        }
    }
}
