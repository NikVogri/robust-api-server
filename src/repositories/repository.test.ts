import { describe, expect, it, vi } from 'vitest';
import { Postgres } from '../services/database/postgres';
import { Logger } from '../services/logging/logger';
import { Repository } from './repository';
import { AppError } from '../error/AppError';

vi.mock('../services/logging/logger');
vi.mock('../services/database/postgres', () => {
    return {
        Postgres: vi.fn(() => ({
            pool: {
                query: vi.fn(),
            },
        })),
    };
});

class TestRepo extends Repository<{ id: string }> {
    constructor(pg: Postgres) {
        super('table-name', pg);
    }
}

describe('Test abstract Repository', () => {
    const logger = new Logger();
    const pg = new Postgres(logger);

    const testRepo = new TestRepo(pg);

    it('should get all the records for table', async () => {
        expect.assertions(2);

        vi.spyOn(pg.pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({
                rows: [{ id: 1 }, { id: 2 }],
            });
        });

        const res = await testRepo.getAll();
        expect(res).toMatchSnapshot();
    });

    it('should should throw an app error if get all the records query failed', async () => {
        expect.assertions(2);

        vi.spyOn(pg.pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            throw new Error('Something went wrong');
        });

        try {
            await testRepo.getAll();
        } catch (err) {
            expect(err).toBeInstanceOf(AppError);
        }
    });

    it('should get records by id', async () => {
        expect.assertions(2);

        vi.spyOn(pg.pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({
                rows: [{ id: 1 }, { id: 2 }],
            });
        });

        const res = await testRepo.getById('1');
        expect(res).toMatchSnapshot();
    });

    it('should should throw an app error if getting a record by id query failed', async () => {
        expect.assertions(2);

        vi.spyOn(pg.pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            throw new Error('Something went wrong');
        });

        try {
            await testRepo.getById('1');
        } catch (err) {
            expect(err).toBeInstanceOf(AppError);
        }
    });

    it('should delete records by id', async () => {
        expect.assertions(2);

        vi.spyOn(pg.pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve();
        });

        const res = await testRepo.delete('1');
        expect(res).toMatchSnapshot();
    });

    it('should should throw an app error if delete record by id query failed', async () => {
        expect.assertions(2);

        vi.spyOn(pg.pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            throw new Error('Something went wrong');
        });

        try {
            await testRepo.delete('1');
        } catch (err) {
            expect(err).toBeInstanceOf(AppError);
        }
    });
});
