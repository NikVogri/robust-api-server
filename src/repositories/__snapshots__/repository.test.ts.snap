// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test abstract Repository > should delete records by id 1`] = `
[
  "DELETE FROM table-name WHERE id = $1",
  [
    "1",
  ],
]
`;

exports[`Test abstract Repository > should delete records by id 2`] = `undefined`;

exports[`Test abstract Repository > should get all the records for table 1`] = `
[
  "SELECT * FROM table-name",
]
`;

exports[`Test abstract Repository > should get all the records for table 2`] = `
[
  {
    "id": 1,
  },
  {
    "id": 2,
  },
]
`;

exports[`Test abstract Repository > should get records by id 1`] = `
[
  "SELECT * FROM table-name WHERE id = $1",
  [
    "1",
  ],
]
`;

exports[`Test abstract Repository > should get records by id 2`] = `
{
  "id": 1,
}
`;

exports[`Test abstract Repository > should should throw an app error if delete record by id query failed 1`] = `
[
  "DELETE FROM table-name WHERE id = $1",
  [
    "1",
  ],
]
`;

exports[`Test abstract Repository > should should throw an app error if get all the records query failed 1`] = `
[
  "SELECT * FROM table-name",
]
`;

exports[`Test abstract Repository > should should throw an app error if getting a record by id query failed 1`] = `
[
  "SELECT * FROM table-name WHERE id = $1",
  [
    "1",
  ],
]
`;
